Bio/SeqUtils/__init__.py

```diff
@@ -667,6 +668,66 @@ def calculate(self, sequence):
 
         return exp(cai_value / cai_length)
 
+    def optimize(self, sequence, seq_type="DNA", strict=False):
+        """Return a new DNA sequence with preferred codons only.
+
+        Uses the codon adaptiveness table defined by the CodonAdaptationIndex
+        object to generate DNA sequences with only preferred codons.
+        May be useful when designing DNA sequences for transgenic protein
+        expression or codon-optimized proteins like fluorophores.
+
+        Arguments:
+            - sequence: DNA, RNA, or protein sequence to codon-optimize.
+                        Supplied as a str, Seq, or SeqRecord object.
+            - seq_type: String specifying type of sequence provided.
+                        Options are "DNA", "RNA", and "protein". Default is "DNA".
+            - strict:   Determines whether an exception should be raised when
+                        two codons are equally prefered for a given amino acid.
+        Returns:
+            Seq object with DNA encoding the same protein as the sequence argument,
+            but using only preferred codons as defined by the codon adaptation index.
+            If multiple codons are equally preferred, a warning is issued
+            and one codon is chosen for use in the optimzed sequence.
+        """
+        try:  # If seq record is provided, convert to sequence
+            sequence = sequence.seq
+        except AttributeError:  # not a  SeqRecord object
+            pass
+        seq = sequence.upper()
+        # Make dict with amino acids referencing preferred codons
+        pref_codons = {}
+        for codon, aminoacid in self._table.forward_table.items():
+            if self[codon] == 1.0:
+                if aminoacid in pref_codons:
+                    if strict:
+                        raise Exception(
+                            f"{pref_codons[aminoacid]} and \
+                                        {codon} are equally preferred."
+                        )
+                    else:
+                        message = f"{pref_codons[aminoacid]} and {codon} \
+                                    are equally preferred. Using {codon}"
+                        warnings.warn(message, RuntimeWarning)
+                pref_codons[aminoacid] = codon
```

Comment:
The backslash inside the string puts a lot of whitespace in the error message. I would probably do something like this:
```py
                    msg = f"{pref_codons[aminoacid]} and {codon} are equally preferred."
                    if strict:
                        raise Exception(msg)
                    warnings.warn(f"{msg} Using {codon}", RuntimeWarning)
```

Reply:
Good catch. You can also do something like this (double check with black), using implicit string concatenation:

```python
                        message = (
                            f"{pref_codons[aminoacid]} and {codon} "
                            f"are equally preferred. Using {codon}"
                        )
```

Biopython as a whole is not consistent about having a trailing space at the end, versus a leading space at the continuing line.